src/plugins/discover/public/application/apps/main/services/use_saved_search.ts

```diff
@@ -0,0 +1,328 @@
+     * because not 100% sure in this case
+     */
+    fieldCounts: Record<string, number>;
+  }>({
+    appState: state,
+    fetchCounter: 0,
+    fieldCounts: {},
+    fetchStatus: shouldSearchOnPageLoad() ? fetchStatuses.LOADING : fetchStatuses.UNINITIALIZED,
+  });
+
+  const fetchAll = useCallback(
+    (reset = false) => {
+      if (!validateTimeRange(timefilter.getTime(), services.toastNotifications)) {
+        return Promise.reject();
+      }
+      const inspectorAdapters = { requests: new RequestAdapter() };
+
+      if (refs.current.abortController) refs.current.abortController.abort();
+      refs.current.abortController = new AbortController();
+
+      // Let the UI know, data fetching started
+      const loadingMessage: SavedSearchSubjectMessage = {
+        state: fetchStatuses.LOADING,
+        fetchCounter: ++refs.current.fetchCounter,
+      };
+
+      if (reset) {
+        // when runtime field was added, changed, deleted, index pattern was switched
+        loadingMessage.rows = [];
+        loadingMessage.fieldCounts = {};
+      }
+      savedSearch$.next(loadingMessage);
+      refs.current.fetchStatus = loadingMessage.state;
+
+      const { sort } = stateContainer.appStateContainer.getState();
+      updateSearchSource({
+        volatileSearchSource: searchSource,
+        indexPattern,
+        services,
+        sort: sort as SortOrder[],
+        useNewFieldsApi,
+      });
+      const chartAggConfigs =
+        indexPattern.timeFieldName && !state.hideChart && state.interval
+          ? getChartAggConfigs(searchSource, state.interval, data)
+          : undefined;
+
+      if (!chartAggConfigs) {
+        searchSource.removeField('aggs');
+      } else {
+        searchSource.setField('aggs', chartAggConfigs.toDsl());
+      }
+
+      const searchSourceFetch$ = searchSource.fetch$({
+        abortSignal: refs.current.abortController.signal,
+        sessionId: searchSessionManager.getNextSearchSessionId(),
+        inspector: {
+          adapter: inspectorAdapters.requests,
+          title: i18n.translate('discover.inspectorRequestDataTitle', {
+            defaultMessage: 'data',
+          }),
+          description: i18n.translate('discover.inspectorRequestDescriptionDocument', {
+            defaultMessage: 'This request queries Elasticsearch to fetch the data for the search.',
+          }),
+        },
+      });
+
+      searchSourceFetch$.subscribe(
+        (res) => {
+          if (!isCompleteResponse(res)) {
+            return;
+          }
+          const documents = res.rawResponse.hits.hits;
+
+          const message: SavedSearchSubjectMessage = {
+            state: fetchStatuses.COMPLETE,
+            rows: documents,
+            inspectorAdapters,
+            fieldCounts: calcFieldCounts(
+              reset ? {} : refs.current.fieldCounts,
+              documents,
+              indexPattern
+            ),
```

Comment:
@majagrubic @timroes with all the changes, the switch to fields, how unmapped fields are handled ... do we still need the implementation of the field counter, that merges the previous result of the fields counts with a new one? this is how it behaved in discover.js, removing it would make the it simpler, but if there are fields that are not in the index pattern mapping, but in the previous result, those would get lost

Reply:
discussed, and postponed this for a future cleanup